staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -799,30 +842,43 @@ func (v *specStandardValidatorV3) validate(schema *apiextensions.JSONSchemaProps
 			if s, err := structuralschema.NewStructural(schema); err == nil {
 				// ignore errors here locally. They will show up for the root of the schema.
 
-				clone := runtime.DeepCopyJSONValue(interface{}(*schema.Default))
-				if !v.isInsideResourceMeta {
-					// If we are under metadata, there are implicitly specified fields like kind, apiVersion, metadata, labels.
-					// We cannot prune as they are pruned as well. This allows more defaults than we would like to.
-					// TODO: be precise about pruning under metadata
-					pruning.Prune(clone, s, s.XEmbeddedResource)
+				validator := govalidate.NewSchemaValidator(s.ToGoOpenAPI(), nil, "", strfmt.Default)
 
-					// TODO: coerce correctly if we are not at the object root, but somewhere below.
-					if err := schemaobjectmeta.Coerce(fldPath, clone, s, s.XEmbeddedResource, false); err != nil {
-						allErrs = append(allErrs, err)
+				if v.isInsideResourceMeta && defaultTester == nil {
+					// this should never happen. But if it does, it's better to fail gracefully than panic
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("default"), "internal validation error"))
+				} else if v.isInsideResourceMeta {
+					// we are inside metadata. We have to construct a test object and plug in the default value to call the
+					// ObjectMeta coercion on it.
+					tester, err := defaultTester(runtime.DeepCopyJSONValue(interface{}(*schema.Default)))
+					if err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Child("default"), interface{}(*schema.Default), err.Error()))
+					} else if err := schemaobjectmeta.Coerce(nil, tester, nil, true, false); err != nil {
+						allErrs = append(allErrs, field.Invalid(fldPath.Child("default"), interface{}(*schema.Default), err.Error()))
+					} else if errs := schemaobjectmeta.Validate(nil, tester, nil, true); len(errs) > 0 {
+						allErrs = append(allErrs, field.Invalid(fldPath.Child("default"), interface{}(*schema.Default), errs.ToAggregate().Error()))
+					} else if errs := apiservervalidation.ValidateCustomResource(fldPath.Child("default"), interface{}(*schema.Default), validator); len(errs) > 0 {
+						allErrs = append(allErrs, errs...)
 					}
-
-					if !reflect.DeepEqual(clone, interface{}(*schema.Default)) {
+				} else {
+					// prune everything other than TypeMeta and ObjectMeta. ObjectMeta will be pruned during storage creation,
+					// but we have to ignore unpruned ObjectMeta values here for API backwards/forward compatibility.
+					pruned := runtime.DeepCopyJSONValue(interface{}(*schema.Default))
+					pruning.Prune(pruned, s, s.XEmbeddedResource)
```

Comment:
@liggitt can you construct something with `x-preserve-unknown-fields: true` in the sourounding schema (i.e. not in `s` itself) that applies to `pruned`?

I think I cannot because pruning only applies to `type: object` nodes and for them we set `x-preserve-unknown-fields` back to false.

Reply:
correct